proxyd/consensus_poller.go

```diff
@@ -0,0 +1,345 @@
+	if lowestBlock == "" {
+		return
+	}
+
+	proposedBlock := lowestBlock
+	proposedBlockHash := lowestBlockHash
+	hasConsensus := false
+
+	// check if everybody agrees on the same block hash
+	consensusBackends := make([]*Backend, 0, len(cp.backendGroup.Backends))
+	consensusBackendsNames := make([]string, 0, len(cp.backendGroup.Backends))
+	filteredBackendsNames := make([]string, 0, len(cp.backendGroup.Backends))
+
+	if lowestBlock > currentConsensusBlockNumber {
+		log.Info("validating consensus on block", lowestBlock)
+	}
+
+	broken := false
+	for !hasConsensus {
+		allAgreed := true
+		consensusBackends = consensusBackends[:0]
+		filteredBackendsNames = filteredBackendsNames[:0]
+		for _, be := range cp.backendGroup.Backends {
+			if be.IsRateLimited() || !be.Online() || time.Now().Before(cp.backendState[be].bannedUntil) {
+				filteredBackendsNames = append(filteredBackendsNames, be.Name)
+				continue
+			}
+
+			actualBlockNumber, actualBlockHash, err := cp.fetchBlock(ctx, be, proposedBlock)
+			if err != nil {
+				log.Warn("error updating backend", "name", be.Name, "err", err)
+				continue
+			}
+			if proposedBlockHash == "" {
+				proposedBlockHash = actualBlockHash
+			}
+			blocksDontMatch := (actualBlockNumber != proposedBlock) || (actualBlockHash != proposedBlockHash)
+			if blocksDontMatch {
+				if blockAheadOrEqual(currentConsensusBlockNumber, actualBlockNumber) {
+					log.Warn("backend broke consensus", "name", be.Name, "blockNum", actualBlockNumber, "proposedBlockNum", proposedBlock, "blockHash", actualBlockHash, "proposedBlockHash", proposedBlockHash)
+					broken = true
+				}
+				allAgreed = false
+				break
+			}
+			consensusBackends = append(consensusBackends, be)
+			consensusBackendsNames = append(consensusBackendsNames, be.Name)
+		}
+		if allAgreed {
+			hasConsensus = true
+		} else {
+			// walk one block behind and try again
+			proposedBlock = hexAdd(proposedBlock, -1)
+			proposedBlockHash = ""
+			log.Info("no consensus, now trying", "block:", proposedBlock)
+		}
+	}
+
+	if broken {
+		// propagate event to other interested parts, such as cache invalidator
+		log.Info("consensus broken", "currentConsensusBlockNumber", currentConsensusBlockNumber, "proposedBlock", proposedBlock, "proposedBlockHash", proposedBlockHash)
+	}
+
+	cp.tracker.SetConsensusBlockNumber(proposedBlock)
+	consensusLatestBlock.Set(blockToFloat(proposedBlock))
+	cp.consensusGroupMux.Lock()
+	cp.consensusGroup = consensusBackends
+	cp.consensusGroupMux.Unlock()
+
+	log.Info("group state", "proposedBlock", proposedBlock, "consensusBackends", strings.Join(consensusBackendsNames, ", "), "filteredBackends", strings.Join(filteredBackendsNames, ", "))
+}
+
+// fetchBlock Convenient wrapper to make a request to get a block directly from the backend
+func (cp *ConsensusPoller) fetchBlock(ctx context.Context, be *Backend, block string) (blockNumber string, blockHash string, err error) {
+	var rpcRes RPCRes
```

Comment:
If you define a struct like:

```go
type ConsensusBlock struct {
  Number hexutil.Uint64 `json:"number"`
  Hash hexutil.Uint64 `json:"hash"`
}
```

Then pass in:

```go
rpcRes := RPCRes{
  Result: ConsensusBlock{}
}
```

You can access the block number as `hexutil.Uint64` directly and avoid having to do any manual conversions between the hex-encoded block number and `uint64`. This would also obviate the need for the manual casting starting on line 302 below, and perform error handling for you.

Reply:
```
rpcRes := RPCRes{
  Result: ConsensusBlock{}
}
```

The `Result` field type is an `interface{}` .  When the result is being unmarshalled, the runtime will unmarshall the result to match it, creating a `map[string]interface{}` from the JSON, regardless of the current value of the `Result` field.

Reply:
The block hash does not fit into a Uint64, keeping as a string

Reply:
`hash` won't fit into a `hexutil.Uint64`, keeping as a string

Reply:
I made all block numbers `hexutil.Uint64` and get rid of hex arithmetic and extra methods